packages/actor-init-sparql/lib/HttpServiceSparqlEndpoint.ts

```diff
@@ -0,0 +1,218 @@
+import * as http from "http";
+import * as querystring from "querystring";
+import {Writable} from "stream";
+import * as url from "url";
+import {newEngineDynamic} from "../index";
+import {ActorInitSparql} from "./ActorInitSparql";
+import {IQueryOptions} from "./QueryDynamic";
+import EventEmitter = NodeJS.EventEmitter;
+import minimist = require("minimist");
+import * as fs from "fs";
+import {LoggerPretty} from "../../logger-pretty";
+
+/**
+ * An HTTP service that exposes a Comunica engine as a SPARQL endpoint.
+ */
+export class HttpServiceSparqlEndpoint {
+  private static readonly MIME_PLAIN = 'text/plain';
+  private static readonly MIME_JSON  = 'application/json';
+
+  private readonly context: any;
+  private readonly timeout: number;
+  private readonly port: number;
+
+  private readonly engine: Promise<ActorInitSparql>;
+
+  constructor(args?: IHttpServiceSparqlEndpointArgs) {
+    args = args || {};
+    this.context = args.context || {};
+    this.timeout = args.timeout || 60000;
+    this.port = args.port || 3000;
+
+    this.engine = newEngineDynamic(args);
+  }
+
+  public static runArgsInProcess(moduleRootPath: string, defaultConfigPath: string) {
+    const args = minimist(process.argv.slice(2));
+    if (args._.length !== 1 || args.h || args.help) {
+      process.stderr.write(
+        'usage: comunica-sparql-http context [-p port] [-t timeout] [-l log-level] [--help]\n' +
+        '  context should be a JSON object, e.g.\n' +
+        '      { "sources": [{ "type": "hypermedia", "value" : "http://fragments.dbpedia.org/2015/en" }]}\n' +
+        '  or the path to such a JSON file\n',
+      );
+      process.exit(1);
+    }
+
+    // allow both files as direct JSON objects for context
+    const context = JSON.parse(fs.existsSync(args._[0]) ? fs.readFileSync(args._[0], 'utf8') : args._[0]);
+    const timeout = (parseInt(args.t, 10) || 60) * 1000;
+    const port = parseInt(args.p, 10) || 3000;
+
+    // Set the logger
+    if (!context.log) {
+      context.log = new LoggerPretty({ level: args.l || 'warn' });
+    }
+
+    const options = {
+      configResourceUrl: process.env.COMUNICA_CONFIG
+        ? process.cwd() + '/' + process.env.COMUNICA_CONFIG : defaultConfigPath,
+      context,
+      mainModulePath: moduleRootPath,
+      port,
+      timeout,
```

Comment:
Just wondering, why are these extra parameters now added to the options object?

Reply:
Because `IHttpServiceSparqlEndpointArgs ` extends from `IQueryOptions `. I wanted to have only a single args object :-)